### PR TITLE
[mono][aot] Emit a list of exported methods into the AOT image and load them when the image is loaded.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

File: src/mono/mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -5303,6 +5306,8 @@ MONO_RESTORE_WARNING
 						export_name = (char *)g_malloc (slen + 1);
 						sprintf (export_name, "%s%s", acfg->user_symbol_prefix, named);
 						export_name [slen] = 0;
+
+						g_ptr_array_add (acfg->exported_methods, method);
```
Comment (kotlarmilos) [inline src/mono/mono/mini/aot-compiler.c:5310]: Can it create duplicate items in an array?

File: src/mono/mono/mini/aot-compiler.c [hunk 2]
```diff
@@ -10874,6 +10879,24 @@ emit_method_info_table (MonoAotCompile *acfg)
...
+	 if (acfg->aot_opts.runtime_init_callback) {
+		 acfg->n_exported_methods = acfg->exported_methods->len;
+		 if (acfg->exported_methods->len) {
+			 guint32 *arr = g_new0 (guint32, acfg->exported_methods->len);
+			 for (i = 0; i < acfg->exported_methods->len; ++i) {
+				 MonoMethod *m = (MonoMethod*)g_ptr_array_index (acfg->exported_methods, i);
+				 arr [i] = mono_method_get_token (m);
+			 }
+			 acfg->exported_methods_offset = add_to_blob_aligned (acfg, (guint8*)arr, acfg->exported_methods->len * sizeof (guint32), 4);
+			 g_free (arr);
```
Comment (kotlarmilos) [inline src/mono/mono/mini/aot-compiler.c:10897]: Should `acfg->exported_methods` be freed here?

File: src/mono/mono/mini/aot-compiler.c (hunk 1 above)
Comment (vargaz) [reply src/mono/mono/mini/aot-compiler.c:5310]: Maybe if the method has multiple unmanagedcallersonly attributes, but it shouldn't be a problem in practice.

**Comment by lateralusX** [comment]
I will run this on the library builder scenario during the day.

**Review by lateralusX — APPROVED** [review]
LGTM! Ran it in both Full AOT and Full AOT + LLVM mode in a scenario where all is build into a native library, unmanagedcallersonly are exported out as entry points, library embedded in application that just call entry point, library init callback gets called that just loads the assembly and that now triggers the load of the methods marked with unmanagedcallersonly attribute, making sure all GOT and PLT slots have been initialized before continue executing the code in native-to-managed wrapper.

**Comment by vargaz** [comment]
Failures are unrelated.

**Pull request merged**